index.md

```diff
@@ -2,9 +2,286 @@
+
+To allow arbitrary scaling we will deploy multiple server instances with some form of load balancing.
+We are designing the server to be stateless such that any client can talk to any server at any time without having a client pinned to a specific server.
+
+#### Multiple Database Servers
+
+Deploying multiple Butler services allows client requests to be handled at scale but that needs to be matched by database capacity.
+Multiple read-only clones of a data release can scale relatively easily.
+There is a complication when clients are allowed to store data in the butler in terms of replication to replicas, and this is discussed later.
+
+#### Query System
+
+Many Butler queries can take minutes or even longer, and it is not desirable to lock up resources in the server process whilst waiting for these queries to complete.
+The solution is to use a system where queries are sent to workers in an execution pool.
+The client will be issued a job ID and can query the server to obtain the worker status.
+Once a job completes the client will receive [question: are we using keep-alive web sockets or polling?] a URI to the results in either JSON or parquet format.
+These results will be written with an expiry date to allow for automatic cleanup of results that are no longer needed.
+If the query resulted in many results they can be written as multiple files and multiple URIs can be returned to allow the client to paginate.
+
+### Writing to a Butler Registry
+
+The biggest complication in implementing client/server Butler is handling writes.
+Clients will never need to update dimension records (those are created by instrument registration, raw data ingest, or visit definitions, which are generally thought of as administrative tasks that can, for now, use direct butler connections) but Data Rights holders will want to read datasets, construct new datasets derived from them, and store them back in a butler.
+
+We assume that there will be far fewer writes than there will be reads and most of the writes will involve multi-dataset transfers from personal workspaces or from user-batch outputs rather than people calling `butler.put()` from notebooks.
+
+There is still some debate over where user-generated products will be stored and that is discussed [later in this document](#writing-datasets).
+
+We have considered multiple options.
+
+#### Use primary database
+
+In this scenario, whilst queries can be spread across multiple database servers, writes always go to a single database server.
+Those updates are then replicated to the other servers, albeit with no immediate consistency guarantees and the possibility that someone could put a dataset and not be able to retrieve it straightaway due to replication latency.
+
+This is the simplest approach and closely matches how writes are handled in the direct butler.
+
+#### Use Butler Workspaces
+
+An option explored in {cite:p}`DMTN-271` is to use the concept of butler workspaces.
```

Comment:
The DMTN-271 workspace concept is actually for limited-butler pseudo-reposoitories that are backed by (e.g.) QuantumGraphs, not true SQLite-backed data repositories.  I'd prefer to avoid using that technote or the "workspace" terminology here (we can just call them personal SQLite/POSIX repos, as you mostly already do anyway).